File: rxhttp/src/main/java/rxhttp/wrapper/parse/MapParser.java
package rxhttp.wrapper.parse;


import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Map;

import okhttp3.Response;
import rxhttp.wrapper.entity.ParameterizedTypeImpl;
import rxhttp.wrapper.exception.ParseException;
import rxhttp.wrapper.utils.GsonUtil;
import rxhttp.wrapper.utils.TypeUtil;

/**
 * 将Response对象解析成泛型Map<K,V>对象
 * User: ljx
 * Date: 2018/10/23
 * Time: 13:49
 */
public class MapParser<K, V> implements Parser<Map<K, V>> {
    private Type kType, vType;

    protected MapParser() {
        kType = TypeUtil.getActualTypeParameter(this.getClass(), 0);
        kType = TypeUtil.getActualTypeParameter(this.getClass(), 1);
    }

    private MapParser(Type kType, Type vType) {
        this.kType = kType;
        this.vType = vType;
    }

    public static <K, V> MapParser<K, V> get(Class<K> kType, Class<V> vType) {
        return new MapParser<>(kType, vType);
    }

    @Override
    public Map<K, V> onParse(Response response) throws IOException {
        String content = getResult(response);
        final Type type = ParameterizedTypeImpl.getParameterized(Map.class, kType, vType);
        Map<K, V> t = GsonUtil.getObject(content, type);  //json 转 对象
        if (t == null)  //解析失败，抛出异常
            throw new ParseException("data parse error");
        return t;
    }
}

File: rxhttp/src/main/java/rxhttp/wrapper/parse/MapParser.kt
package rxhttp.wrapper.parse

import okhttp3.Response
import rxhttp.wrapper.entity.ParameterizedTypeImpl
import rxhttp.wrapper.utils.TypeUtil
import java.io.IOException
import java.lang.reflect.Type

/**
 * 将Response对象解析成泛型Map对象
 * User: ljx
 * Date: 2018/10/23
 * Time: 13:49
 */
open class MapParser<K, V> : Parser<MutableMap<K, V>> {
    private var kType: Type
    private var vType: Type

    protected constructor() {
        kType = TypeUtil.getActualTypeParameter(this.javaClass, 0)
        vType = TypeUtil.getActualTypeParameter(this.javaClass, 1)
    }

    constructor(kType: Class<K>, vType: Class<V>) {
        this.kType = kType
        this.vType = vType
    }

    @Throws(IOException::class)
    override fun onParse(response: Response): MutableMap<K, V> {
        val type: Type = ParameterizedTypeImpl.getParameterized(MutableMap::class.java, kType, vType)
        return convert(response, type)
    }

    companion object {
        @JvmStatic
        operator fun <K, V> get(kType: Class<K>, vType: Class<V>) = MapParser(kType, vType)
    }
}